File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.py
import sys

N = int(input())
Nlist = list(map(int,sys.stdin.readline().split()))

realanswer = 0
visited= [False] * N

def DFS(answerarr,depth) :

    global realanswer
    global visited

    if depth == N :
        #다돌았다.
        answer =0 
        for i in range(N-1):
            answer += abs(answerarr[i]-answerarr[i+1])
        realanswer = max(answer,realanswer)
        return

    for i in range(len(Nlist)):
        if visited[i] == False :
            visited[i] = True
            answerarr.append(Nlist[i])
            DFS(answerarr,depth+1)
            visited[i] = False
            answerarr.pop()
            
                
DFS([],0)

print(realanswer)

File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    static int N ; 
    static int [] originarr ;
    static boolean [] visited;
    static int answer = Integer.MIN_VALUE;
    
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());
        originarr = new int [N];
         String [] oneline = br.readLine().split(" ");
        for(int j = 0 ; j<N ; j++)
            {
                originarr[j] = Integer.parseInt(oneline[j]);
            }
        visited = new boolean[N];

        for (int i = 0 ; i<N ;i++)
            {
                List<Integer> newlist = new LinkedList<>();
                visited[i] = true;
                newlist.add(originarr[i]);
                
                DFS(1,newlist);
                
                visited[i] = false;
                
            }

        bw.write(String.valueOf(answer));
        bw.close();
    }

    public static void DFS(int depth, List<Integer> newlist)
    {
        if(depth==N)
        {
            int cnt = 0;
            for (int i = 1 ; i<N; i++)
                {
                    cnt += Math.abs(newlist.get(i)-newlist.get(i-1));
                }
            answer = Math.max(answer,cnt);
            return;
        }

        for (int i = 0 ; i<N ; i++)
            {
                if(!visited[i])
                {
                    visited[i] = true;
                    newlist.add(originarr[i]);
                    DFS(depth+1,newlist);
                    newlist.remove(newlist.size()-1);
                    visited[i] = false;
                }
            }
        return;
    }
}